modules/core/test/src/smithy4s/schema/SchemaPartitionSpec.scala

```diff
@@ -0,0 +1,222 @@
+      int.required[Foo]("y", _.y)
+    )(Foo.apply)
+
+    // We're splitting the schema into various components. The first one is extracted
+    // as a "payload" partial, which means that schema held by the first field
+    // matching the predicate will be used as if it was a top level schema
+    val xPartialSchema = schema.payloadPartition(_.label == "x")
+
+    val documentX = Document.array(Document.fromInt(1), Document.fromInt(2))
+    val documentY = Document.obj("y" -> Document.fromInt(1))
+
+    import SchemaPartition._
+    xPartialSchema match {
+      case SplittingMatch(xSchema, ySchema) =>
+        val decoderX = Document.Decoder.fromSchema(xSchema)
+        val decoderY = Document.Decoder.fromSchema(ySchema)
+
+        val result =
+          (decoderX.decode(documentX), decoderY.decode(documentY)).mapN {
+            PartialData.unsafeReconcile(_, _)
+          }
+
+        assertEquals(result, Right(Foo(List(1, 2), 1)))
+
+      case _ => fail("Expected partial matches")
+    }
+  }
+
+  test(
+    "Computing a partial schema using a predicate that matches all fields returns the original schema"
+  ) {
+    case class Foo(x: Int, y: Option[Int])
+    val schema = struct(
+      int.required[Foo]("x", _.x),
+      int.optional[Foo]("y", _.y)
+    )(Foo.apply)
+
+    val partialSchema = schema.partition(_ => true)
+    assert(partialSchema == SchemaPartition.TotalMatch(schema))
+  }
+
+  test(
+    "Computing a partial schema using a predicate that matches no field returns no schema"
+  ) {
+    case class Foo(x: Int, y: Option[Int])
+    val schema = struct(
+      int.required[Foo]("x", _.x),
+      int.optional[Foo]("y", _.y)
+    )(Foo.apply)
+
+    val partialSchema = schema.partition(_.label.startsWith("z"))
+    assertEquals(partialSchema, SchemaPartition.NoMatch[Foo]())
+  }
+
+  test(
+    "Computing a partial schema from a non-struct schema returns no-schema"
+  ) {
+    val schema = list(int)
+
+    val partialSchema = schema.partition(_.label.startsWith("z"))
+    assertEquals(partialSchema, SchemaPartition.NoMatch[List[Int]]())
+  }
+
+  test(
+    "Payload-specialised partial schemas are considered total when the original schema had a single field"
+  ) {
+    case class Foo(x: List[Int])
+    val schema = struct(
+      list(int).required[Foo]("x", _.x)
+    )(Foo.apply)
+
+    // We're splitting the schema into various components. The first one is extracted
+    // as a "payload" partial, which means that schema held by the first field
+    // matching the predicate will be used as if it was a top level schema
+    val xPartialSchema = schema.payloadPartition(_.label == "x")
+
+    val documentX = Document.array(Document.fromInt(1), Document.fromInt(2))
+
+    import SchemaPartition._
+    xPartialSchema match {
+      case (TotalMatch(xSchema)) =>
+        val decoderX = Document.Decoder.fromSchema(xSchema)
+
+        val result = decoderX.decode(documentX)
+
+        assert(result == Right(Foo(List(1, 2))))
```

Comment:
**thought:** I would love to see this defined in a more "property" style, like

```scala
val decoderX = Document.Decoder.fromSchema(xSchema)
val decoderBase = Document.Decoder.fromSchema(schema)

assertEquals(
  decoderX.decode(documentX),
  decoderBase.decode(
    Document.obj(
      "x" -> documentX
    )
  )
)
```

then it becomes a bit clearer how the `xSchema` actually differs from the original schema we started with

Reply:
Good idea